util/src/main/java/tech/pegasys/artemis/util/backing/tree/TreeNode.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.backing.tree;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.function.Function;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.crypto.Hash;
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.artemis.util.backing.tree.TreeNodeImpl.RootImpl;
+
+/**
+ * Basic interface for Backing Tree node Backing Binary Tree concept for SSZ structures is described
+ * here: https://github.com/protolambda/eth-merkle-trees/blob/master/typing_partials.md#tree
+ *
+ * <p>Tree node is immutable by design. Any update on a tree creates new nodes which refer both new
+ * data nodes and old unmodified nodes
+ */
+public interface TreeNode {
+
+  static TreeNode createRoot(Bytes32 val) {
+    return new RootImpl(val);
+  }
+
+  /** Leaf node of a tree which contains 'bytes32' value */
+  interface Root extends TreeNode {
```

Comment:
I find the naming of these types a bit confusing.  What about:
```suggestion
  interface LeafNode extends TreeNode {
```

Reply:
I'm also not too comfortable with `Root` and `Commit` terms, but this is exactly how they are defined in the spec: https://github.com/protolambda/eth-merkle-trees/blob/master/typing_partials.md#structure
I would prefer to stick with spec to avoid confusion, since the javadoc refers to this spec in several places.

Reply:
I'd prefer clarifying the java names, and making a note that "LeafNode" corresponds to "Root" in the linked spec, etc.  But happy to leave the decision with you.

Reply:
Let's rename them. I remember `Root` class name confused me a couple of times during development

Reply:
https://github.com/PegaSysEng/teku/pull/1133/commits/dc7100825b3a7dd362f957c75883bb933983aee0